### PR TITLE
add windows-installer example pipeline

### PR DESCRIPTION
**Pull request by suomiy**
Downloads a Windows Source ISO into a PVC and installs Windows into a new root PVC.
It uses Windows Sysprep to automatically install the system. 
The pipeline can be customized to support different installation requirements. 

The artifact of this pipeline is a PVC (backed by DataVolume) with a custom Windows installation. 

[README](https://github.com/suomiy/kubevirt-tekton-tasks/tree/win-pipeline/examples/pipelines/windows-installer)

Todo:
- [x] implement wait-for-vm task (https://github.com/kubevirt/kubevirt-tekton-tasks/pull/81)

### CODE REVIEW

**Comment by suomiy** [comment]
@dankenigsberg @fabiand

I decided to split the Windows flow into 2 pipelines.

1. This one which installs the windows and has a PVC as an artifact.
2. Pipeline which copies this base PVC and runs a build/deployment from that. 

IMO a single pipeline with these two use cases would be too cumbersome.

Also this pipeline requires a `wait-for-vm` task which will detect that the VM turned off and the installation has completed. 
This task needs to be implemented before this pipeline. 

Any suggestions for a better name than `wait-for-vm`?

File: examples/pipelines/windows-installer/README.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+```bash
+WIN_URL="https://software-download.microsoft.com/db/Win10_20H2_v2_English_x64.iso..."
+kubectl apply -f windows-installer-pipeline.yaml
+sed 's!DOWNLOAD_URL!'"$WIN_URL"'!g' windows-installer-pipelinerun.yaml | kubectl create -f -
+```
+
+## Possible Optimizations
+
+- Windows source ISO can be downloaded and cached/stored in a PVC. 
+  So the subsequent PipelineRuns don't have to download the same ISO multiple times.
```
Comment (dankenigsberg) [inline examples/pipelines/windows-installer/README.md:60]: Can you provide the optimization right now? The pipeline should better work without Internet connectivity.

File: examples/pipelines/windows-installer/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+4. Initialize a WIN_URL variable that will be used to create a DataVolume which will download this ISO into a PVC.
+
+```bash
+# Real URL can look differently
+WIN_URL="https://software-download.microsoft.com/db/Win10_20H2_v2_English_x64.iso..."
+```
+
+### Prepare autounattend.xml ConfigMap
+
+1. Supply, generate or use the default autounattend.xml.
```
Comment (dankenigsberg) [inline examples/pipelines/windows-installer/README.md:28]: I'm not sure how I can generate or where I can supply my own autounattend.xml file. I suppose that a usage example of the suggested Pipeline would help.

**Review by dankenigsberg — CHANGES_REQUESTED** [review]
> I decided to split the Windows flow into 2 pipelines.
> 
>     1. This one which installs the windows and has a PVC as an artifact.

Ok, as this first pipeline seems useful on its own right. I'd love to tweak it so it places its artifact under openshift-virtualization-os-images namespace to be auto-cloned by GUI when a new VM is requested.

>     2. Pipeline which copies this base PVC and runs a build/deployment from that.

Not sure what this consists.
> 
> Also this pipeline requires a `wait-for-vm` task which will detect that the VM turned off and the installation has completed.
> This task needs to be implemented before this pipeline.
> 
> Any suggestions for a better name than `wait-for-vm`?

Should this task wait for a vm to shut down? If so, I'd call it `wait-for-vm-shutdown`. Does Tekton have similar tasks for waiting for pods?

File: examples/pipelines/windows-installer/windows-installer-pipeline.yaml [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+                    <SynchronousCommand wcm:action="add">
+                        <Order>3</Order>
+                        <RequiresUserInput>false</RequiresUserInput>
+                        <CommandLine>cmd /C wmic useraccount where name="Administrator" set PasswordExpires=false</CommandLine>
+                        <Description>Password Never Expires</Description>
+                    </SynchronousCommand>
+                    <SynchronousCommand wcm:action="add">
+                        <Order>4</Order>
+                        <RequiresUserInput>false</RequiresUserInput>
+                        <CommandLine>cmd /c shutdown /s /f /t 10</CommandLine>
```
Comment (eifrach) [inline examples/pipelines/windows-installer/windows-installer-pipeline.yaml:178]: I'm missing here the sysprep command.
you mention it in the readme  but no reference to it

**Comment by suomiy** [comment]
> > I decided to split the Windows flow into 2 pipelines.
> > ```
> > 1. This one which installs the windows and has a PVC as an artifact.
> > ```
> 
> Ok, as this first pipeline seems useful on its own right. I'd love to tweak it so it places its artifact under openshift-virtualization-os-images namespace to be auto-cloned by GUI when a new VM is requested.

Yes, it can be used this way as one of the use case , but we should  keep it runnable on Kubernetes in upstream as well.

> 
> > ```
> > 2. Pipeline which copies this base PVC and runs a build/deployment from that.
> > ```
> 
> Not sure what this consists.

this is TBD

> 
> > Also this pipeline requires a `wait-for-vm` task which will detect that the VM turned off and the installation has completed.
> > This task needs to be implemented before this pipeline.
> > Any suggestions for a better name than `wait-for-vm`?
> 
> Should this task wait for a vm to shut down? If so, I'd call it `wait-for-vm-shutdown`. Does Tekton have similar tasks for waiting for pods?

They have a similar thing for [kubectl-deploy-pod](https://github.com/tektoncd/catalog/tree/main/task/kubectl-deploy-pod/0.1), but I would rather keep it in a separate task since VM's have more complex lifecycle.

They use `success-condition` and `failure-condition`, so we could do something similar to that. I would prefer to have a single task rather than multiple of `wait-for-vm-shutdown`, etc. Since you might want to wait for Running / Succeeded / Failed

File: examples/pipelines/windows-installer/windows-installer-pipeline.yaml (hunk 3 above)
Comment (suomiy) [reply examples/pipelines/windows-installer/windows-installer-pipeline.yaml:178]: It should be run automatically by the windows installer when it locates `autounattend.xml`

File: examples/pipelines/windows-installer/README.md (hunk 2 above)
Comment (suomiy) [reply examples/pipelines/windows-installer/README.md:28]: I can put a link to the Microsoft documenation there if that helps.

File: examples/pipelines/windows-installer/README.md (hunk 1 above)
Comment (suomiy) [reply examples/pipelines/windows-installer/README.md:60]: I would prefer for the pipeline to be as simple as possible so one can run the example quickly.  But I can describe the extra steps here in this chapter.

File: examples/pipelines/windows-installer/README.md (hunk 2 above)
Comment (dankenigsberg) [reply examples/pipelines/windows-installer/README.md:28]: A link would help, but I'm asking for an example of a pipeline with a non-default `autounattended.xml`

**Comment by dankenigsberg** [comment]
> They use success-condition and failure-condition, so we could do something similar to that. I would prefer to have a single task rather than multiple of wait-for-vm-shutdown, etc. Since you might want to wait for Running / Succeeded / Failed

`wait-for-vm-status` then?

File: examples/pipelines/windows-installer/README.md (hunk 2 above)
Comment (suomiy) [reply examples/pipelines/windows-installer/README.md:28]: I am planning to add ssh server installation to this pipeline later. But other use case specific flow can be implemented as an addon / other pipeline later.

**Comment by suomiy** [comment]
> > They use success-condition and failure-condition, so we could do something similar to that. I would prefer to have a single task rather than multiple of wait-for-vm-shutdown, etc. Since you might want to wait for Running / Succeeded / Failed
> 
> `wait-for-vm-status` then?

1. Shouldn't we rather name it `wait-for-vmi-status` since it will be checking VMI status? 
2. It might include checking for VM status in the future, but that is probably too much of an edge case

**Comment by dankenigsberg** [comment]
> 
>     1. Shouldn't we rather name it `wait-for-vmi-status` since it will be checking VMI status?

wfm

> 
>     2. It might include checking for VM status in the future, but that is probably too much of an edge case

I don't understand what you're saying here

**Comment by suomiy** [comment]
> > ```
> > 1. Shouldn't we rather name it `wait-for-vmi-status` since it will be checking VMI status?
> > ```
> 
> wfm
> 
> > ```
> > 2. It might include checking for VM status in the future, but that is probably too much of an edge case
> > ```
> 
> I don't understand what you're saying here

if it should have other detection abilities in the future with regards to this status: https://kubevirt.io/api-reference/master/definitions.html#_v1_virtualmachinestatus

**Comment by suomiy** [comment]
PR for wait-for-vmi-status task: #81

File: examples/pipelines/windows-installer/README.md (hunk 2 above)
Comment (suomiy) [reply examples/pipelines/windows-installer/README.md:28]: fixed

File: examples/pipelines/windows-installer/README.md (hunk 1 above)
Comment (suomiy) [reply examples/pipelines/windows-installer/README.md]: fixed

**Comment by suomiy** [comment]
#81 has been merged and the pipeline should be getting ready for merging

**Comment by suomiy** [comment]
merging for https://github.com/kubevirt/kubevirt.github.io/pull/727

**Pull request merged**